..........**Pipeline**......

by pipelines what we do is to
output become input for the next step....
example  = lets we have to make machinelearnig model but therre is huge problems in dataset like have to filling missing variable and encode categorical variabe , scaling rthe data..... and so on so there is so many steps to process the data ........but by pipelines we could easily to do it .......when we deploy my model into the server then the input which are coming again we have to clean preprocess but through pipelines its becomes quite easy....

**Titanic without using pipeline**

In [3]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.tree import DecisionTreeClassifier


In [4]:
v = pd.read_csv('/content/train.csv.csv.csv')

In [5]:
v.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [6]:
v.drop(columns = ['PassengerId' , 'Name' ,'Ticket' , 'Cabin'] ,inplace = True)

In [7]:
v.head()

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0         0       3    male  22.0      1      0   7.2500        S
1         1       1  female  38.0      1      0  71.2833        C
2         1       3  female  26.0      0      0   7.9250        S
3         1       1  female  35.0      1      0  53.1000        S
4         0       3    male  35.0      0      0   8.0500        S

step 1 train test split

In [8]:
# here we  have droped survived column fromm x axis but
# retained in y axis... because from x axis inputs we have to predict y axis
# and y axis  is  survived column

x_train , x_test , y_train , y_test = train_test_split(v.drop(columns = ['Survived']) , v['Survived'] , test_size = 0.2 , random_state = 42)



In [9]:
x_train.head()

Pclass     Sex   Age  SibSp  Parch     Fare Embarked
331       1    male  45.5      0      0  28.5000        S
733       2    male  23.0      0      0  13.0000        S
382       3    male  32.0      0      0   7.9250        S
704       3    male  26.0      1      0   7.8542        S
813       3  female   6.0      4      2  31.2750        S

starange !!  the sample randomly which our machine choosen ...fot survived coloumn all  are dead

In [10]:
y_train.head()

331    0
733    0
382    0
704    0
813    0
Name: Survived, dtype: int64

In [11]:
v.isnull().sum()

Survived      0
Pclass        0
Sex           0
Age         177
SibSp         0
Parch         0
Fare          0
Embarked      2
dtype: int64

...by null function we could see that age and emabarked have 177 and 2 missing  values respectivey........

In [12]:
# apllying imputation
si_age = SimpleImputer() # made object for missing values for age
si_embarked = SimpleImputer(strategy = 'most_frequent') # made object for
# embarked for missing values ....strategy syntax means in this... replace missig
# values from the mean in embarked


# in this we are fitting tha data means selecting the mising values then
# transform through the simple imputer class ...imputer trandform
# the missing values
x_train_age = si_age.fit_transform(x_train[['Age']])
x_train_embarked = si_embarked.fit_transform(x_train[['Embarked']])
# same for x test too
x_test_age = si_age.transform(x_test[['Age']])
x_test_embarked = si_embarked.transform(x_test[['Embarked']])



** importing one hot encoder class to transform categorical values or strings values into  numbers **

one hot encoding sex and embaked

In [13]:
ohe_sex = OneHotEncoder(sparse_output = False , handle_unknown = 'ignore')
ohe_embarked = OneHotEncoder(sparse_output = False , handle_unknown = 'ignore')

x_train_sex = ohe_sex.fit_transform(x_train[['Sex']])
x_train_embarked = ohe_embarked.fit_transform(x_train_embarked)

x_test_sex = ohe_sex.transform(x_test[['Sex']])
x_test_embarked = ohe_embarked.transform(x_test_embarked)

In [14]:
x_train_embarked

array([[0., 0., 1.],
       [0., 0., 1.],
       [0., 0., 1.],
       ...,
       [0., 0., 1.],
       [0., 0., 1.],
       [0., 0., 1.]])

Here, x_train_embarked becomes a NumPy array,
 because SimpleImputer (and most Scikit-learn functions)
  return NumPy arrays, not DataFrames.

In [15]:
x_train_age

array([[45.5       ],
       [23.        ],
       [32.        ],
       [26.        ],
       [ 6.        ],
       [24.        ],
       [45.        ],
       [29.        ],
       [29.49884615],
       [29.49884615],
       [42.        ],
       [36.        ],
       [33.        ],
       [17.        ],
       [29.        ],
       [50.        ],
       [35.        ],
       [38.        ],
       [34.        ],
       [17.        ],
       [11.        ],
       [61.        ],
       [30.        ],
       [ 7.        ],
       [63.        ],
       [20.        ],
       [29.49884615],
       [29.        ],
       [36.        ],
       [29.49884615],
       [50.        ],
       [27.        ],
       [30.        ],
       [33.        ],
       [29.49884615],
       [29.49884615],
       [ 2.        ],
       [25.        ],
       [51.        ],
       [25.        ],
       [29.49884615],
       [29.49884615],
       [24.        ],
       [18.        ],
       [29.49884615],
       [25

In [16]:
x_train_sex

array([[0., 1.],
       [0., 1.],
       [0., 1.],
       ...,
       [0., 1.],
       [1., 0.],
       [0., 1.]])

In [17]:
x_train.head(2)

Pclass   Sex   Age  SibSp  Parch  Fare Embarked
331       1  male  45.5      0      0  28.5        S
733       2  male  23.0      0      0  13.0        S

so we have three different array of age , sex and embarked ....now  we are gonna call concatenate for others x coulumn....

here what we are gonna do to drop coloumn sex , age and embarked from x train coloumn because it is already is in

we made a variable to store the x train drop syntax .... here we  dropped sex age embarked from the coulumn because we already transformed thse coloumn like imoutation encoding etc etc...
that whyt we dropped ...beacuse id we gain includes then my model could be confuse .....

here from train data  we dropped sex age and columns..

In [18]:
x_train_rem = x_train.drop(columns = ['Sex' , 'Age','Embarked'])

here from test data we removed too those coloumns ....

In [19]:
x_test_rem = x_test.drop(columns = ['Sex' , 'Age','Embarked'])

in below sytax we took x train rem  == means it includes all the coulumn excpt age sex  embarked .... in upper syntax  x train  includes each coulumn except servival ...we dropped because age and other that two coloum  already transformed and became digit if we have not been dropeed from x train then it will be in string ...concatenate...then concatenate will made array of 6 coloumn but there will be string .... that why we dropped ..3 coloumn from the reak data frae and includes other coloumn of x ...then we kept all three transformed  column like age sex embarked in concatenate .....because that is transformed already ....
now concatenate will make x coloumn array off all the 6 coloumn ..

In [20]:
x_train_transformed = np.concatenate((x_train_rem , x_train_age , x_train_sex , x_train_embarked) , axis = 1)
x_test_transformed = np.concatenate((x_test_rem , x_test_age , x_test_sex , x_test_embarked) , axis = 1)

below class we are calling because i am  gonna train  my x coloums model to  predict  y ........clf.fit will learn the pattrens between the x axis columns input and the output y ...it will try to learn the pattern or draw or it will be trained on the pattern ......

here why we have imported decision tree claassifier rather than importing logistic regression because  we have so many numbers of coloumn of x then decision tree good for predicting the output......than logistic regression ...

In [21]:
clf = DecisionTreeClassifier()
clf.fit(x_train_transformed , y_train)

DecisionTreeClassifier()

by below function we give my test input which are not trained clf .predict then it will try  to  figuring out or try to relate  the pattern from test input to trained input then it will give output....

In [22]:
y_pred = clf.predict(x_test_transformed)

In [23]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test , y_pred)*100

78.77094972067039

cant plot decision regions because
 because x  has 6 coloumn or 6 inputs  so plotting fails in 6 d only works till 2 inputs .... because plots works till 2d not above here we have 6 d ....

In [24]:
import pickle

In [25]:
import os
os.makedirs('models', exist_ok=True)

# os is a built-in Python module that helps you interact
# with your computer’s
# Operating System — that’s why it’s called os.
# Create a folder named models if it doesn't already exist."




In [26]:
pickle.dump(ohe_sex ,open('models/ohe_sex.pkl','wb'))
# we have to dump this
# file too because this files contains lots  training data like male is
# female is 0 by this new dat will identify whom to  to keep 0 and whom to
# keep 1 ... by this we could agin train and deploy my model too .

pickle.dump(ohe_embarked,open('models/ohe_embarked.pkl','wb'))

# same case  with embarked too .....
# we are saving my trained model by pickle file

pickle.dump(clf,open('models/clf.pkl','wb'))


#pickle – A Python module used to save Python objects (like models) to a file.

# clf – This is your trained model
# means clf became a object of logistic regression model
# clf  is just a variable name
# (for example, a DecisionTree, LogisticRegression, etc.)

# 'models/clf.pkl' – File path where you want to save the model.

# 'wb' – Write Binary mode – required when saving with pickle.


**so here what we are  goonna do to process and clean the data which are coming from the outside ... then integrate it.. in my old  preprocessed model...**


In [27]:
import pickle
import numpy as np


In [28]:
ohe_sex = pickle.load(open('models/ohe_sex.pkl','rb'))
ohe_embarked = pickle.load(open('models/ohe_embarked.pkl','rb'))
clf = pickle.load(open('models/clf.pkl','rb'))

In [29]:
# Assume user input
# Pclass/gender/size/sibsp/parch/fare/embarked
test_input = np.array([2, 'male', 31.0, 0, 0, 10.5, 'S'],dtype=object).reshape(1,7)




In [30]:
test_input

array([[2, 'male', 31.0, 0, 0, 10.5, 'S']], dtype=object)

In [31]:
test_input_sex = ohe_sex.transform(test_input[:,1].reshape(1,1))

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but OneHotEncoder was fitted with feature names
  warnings.warn(


In [32]:
test_input_sex

array([[0., 1.]])

0,1 indicates for male and 1,0 indicates for female .......


In [33]:
test_input_embarked = ohe_embarked.transform(test_input[:,-1].reshape(1,1))

In [34]:
test_input_embarked

array([[0., 0., 1.]])

In [35]:
test_input_age = ohe_embarked.transform(test_input[:,1].reshape(1,1))

In [36]:
test_input_age = test_input[:,2].reshape(1,1)

In [37]:
test_input_age

array([[31.0]], dtype=object)

In [38]:
test_input_transformed = np.concatenate((test_input[:,[0,3,4,5]] , test_input_age,
    test_input_sex , test_input_embarked) , axis = 1)